src/lsm/tree.zig

```diff
@@ -228,32 +235,119 @@ pub fn TreeType(comptime Table: type) type {
                 if (tree.table_mutable.get(key) orelse
                     tree.value_cache.?.getKeyPtr(tombstone_from_key(key))) |value|
                 {
-                    callback(unwrap_tombstone(value));
+                    callback(context, unwrap_tombstone(value));
                     return;
                 }
             }
 
-            // Hash the key to the fingerprint only once and reuse for all bloom filter checks.
             if (!tree.table_immutable.free and tree.table_immutable.snapshot_min < snapshot) {
                 if (tree.table_immutable.get(key)) |value| {
-                    callback(unwrap_tombstone(value));
+                    callback(context, unwrap_tombstone(value));
                     return;
                 }
             }
 
-            var it = tree.manifest.lookup(snapshot, key);
-            if (it.next()) |info| {
-                assert(info.visible(snapshot));
-                assert(compare_keys(key, info.key_min) != .lt);
-                assert(compare_keys(key, info.key_max) != .gt);
+            // Hash the key to the fingerprint only once and reuse for all bloom filter checks.
+            const fingerprint = bloom_filter.Fingerprint.create(mem.asBytes(&key));
+
+            context.* = .{
+                .tree = tree,
+                .completion = undefined,
+                .fingerprint = fingerprint,
+                .it = tree.manifest.lookup(snapshot, key),
+                .callback = callback,
+            };
 
-                // TODO
-            } else {
-                callback(null);
-                return;
-            }
+            context.read_next_table();
         }
 
+        const LookupContext = struct {
+            tree: *Tree,
+            completion: Grid.Read,
+            fingerprint: bloom_filter.Fingerprint,
+            it: Manifest.LookupIterator,
+
+            current_table: ?struct {
+                data_block_address: u64,
+                data_block_checksum: u128,
+            } = null,
```

Comment:
We could do this with something like `manifest.table_for_key(snapshot, key, level) ?*const TableInfo`, then copying only the checksum and address for each level.

Reply:
Implemented this in e8c99fae7acbd7222064ff4cedc4bd9234f0a1cf. I used the existing manifest iterator in a while loop instead of introducing new manifest API but it would be easy to swap that out if we change the public Manifest API.

Reply:
Great.